### PR TITLE
Precision audit: extend two-layer assertions to fixtures 13-58 + 59

### PR DESCRIPTION
**Pull request by khatchad**
## Summary

Continues the precision-audit pattern landed in #520 (`testHasLikelyTensorParameter11`). Extends the two-layer pattern (per-parameter `getTensorTypes()` + `inferInputSignature()`) to 42 more existing fixtures. All pass against Ariadne 0.44.0 + the current algorithm. No Layer-1 (per-parameter Ariadne) precision gaps found; one notable Layer-2 (algorithm) behavioral discovery at fixture 59 (graceful per-dim wildcard emission for RaggedTensor).

## Coverage In This PR

- **Common-pattern fixtures (mental-model + Ariadne convergence):** 13, 14, 15, 16, 19, 20, 21, 23, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 45, 47, 48, 49, 51, 55, 58
- **Edge-case fixtures (runtime-verified via `python3.10` asserts added to `.py`):** 22 (`tf.Variable`), 24 (`tf.SparseTensor`), 25 (raw `tf.Tensor` constructor)
- **Import-path variants:** 37, 38, 39, 40, 41, 42, 43, 44, 46, 50, 56, 57 (various `tensorflow.python.ops.*`, fully-qualified, etc.)
- **RaggedTensor (new edge case):** 59 — Ariadne emits `(INT32, [3, null, null])` matching runtime; algorithm converts nulls to `SymbolicDim("?")` wildcards in the inferred signature

## Methodology

Per the project audit pattern: read the fixture's Python source, derive the expected per-parameter `TensorType` and `InputSignature` from semantics. For common patterns (`tf.ones`, `tf.zeros`, `tf.constant`), mental-model + Ariadne convergence is sufficient evidence. For edge cases (Variable, SparseTensor, raw Tensor, RaggedTensor), `python3.10 A.py` runs with inline asserts that confirm the runtime ground truth before JUnit assertions are written.

Also static-imports `DType.INT32` (used across fixtures 21, 24, 26, 28, 30, 31, 43-44, 46, 47-48, 50, 57, 59).

## Notable Behavioral Finding (#59)

Ariadne emits `TensorType(INT32, [Constant(3), null, null])` for `tf.RaggedTensor.from_nested_row_splits` — partial-known dims preserve the known constant and emit `null` for unknown rank-2 components. The algorithm then converts `null` dims to `SymbolicDim("?")` wildcards in the inferred signature, demonstrating graceful per-dim degradation rather than a whole-shape drop. Documents an undocumented property of the `inferSpec` reduction.

## Remaining Work (Out of Scope For This PR)

Fixtures 60-158 (~90 tests) use a shared `testHasLikelyTensorParameterHelper()` method, which doesn't currently expose precision-audit parameters. Continuing the audit through that range requires either an overload that accepts an expected `TensorType`, or inlining the helper for each audited test. Both are mechanical follow-ups; leaving for a separate PR to keep this one bounded.

## Test Plan

- [x] `./mvnw verify`: 492 tests, 0 failures, 0 errors, 11 skipped
- [x] `python3.10 A.py` passes for fixtures with edge-case runtime asserts added (22, 24, 25)
- [x] Spotless passes (pre-commit hook)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: edu.cuny.hunter.hybridize.tests/test cases/edu/cuny/hunter/hybridize/tests/HybridizeFunctionRefactoringTest.java [hunk 1]
```diff
@@ -2237,6 +2238,30 @@ public void testHasLikelyTensorParameter13() throws Exception {
 
 			assertTrue("Expecting " + function + " to likely have a tensor-like parameter.", function.getHasTensorParameter());
 		}
+
+		// Precision audit. Top-level call: `dense_layer(tf.ones([3, 2]), tf.ones([2, 2]), tf.ones([2]))`. Inside the body,
+		// `add(tf.matmul(x, w), b)` calls `add` with the matmul result and `b`. Reading the semantics, every tensor here is FLOAT32
+		// with concrete shape: dense_layer's `x` is (3, 2), `w` is (2, 2), `b` is (2,); add's `a` is the matmul output (3, 2), and add's
+		// `b` threads through from dense_layer's `b` so it is also (2,).
+		Function denseLayerFunc = nameToFunctions.get("dense_layer").iterator().next();
```
[the remark this answers is not in the document]
Comment (khatchad) [reply edu.cuny.hunter.hybridize.tests/test cases/edu/cuny/hunter/hybridize/tests/HybridizeFunctionRefactoringTest.java:2246]: `nameToFunctions` here is `Map<String, List<Function>>` from `Collectors.groupingBy(Function::getSimpleName)` (see line 2282). `iterator().next()` on a `List` is deterministic — it returns element 0. The outer loop at lines 2285-2287 also asserts `assertEquals(1, functionList.size())` for every FUT name, so a name collision would fail loudly before reaching the precision-audit block. No action needed.

File: edu.cuny.hunter.hybridize.tests/test cases/edu/cuny/hunter/hybridize/tests/HybridizeFunctionRefactoringTest.java [hunk 2]
```diff
@@ -2237,6 +2238,30 @@ public void testHasLikelyTensorParameter13() throws Exception {
...
+		assertEquals(Set.of(new TensorType(FLOAT32, List.of(new NumericDim(3), new NumericDim(2)))), dlParams.get(0).getTensorTypes());
+		assertEquals(Set.of(new TensorType(FLOAT32, List.of(new NumericDim(2), new NumericDim(2)))), dlParams.get(1).getTensorTypes());
+		assertEquals(Set.of(new TensorType(FLOAT32, List.of(new NumericDim(2)))), dlParams.get(2).getTensorTypes());
+		Optional<InputSignature> dlSig = denseLayerFunc.inferInputSignature();
+		assertTrue(dlSig.isPresent());
+		assertEquals(List.of(new TensorType(FLOAT32, List.of(new NumericDim(3), new NumericDim(2))),
+				new TensorType(FLOAT32, List.of(new NumericDim(2), new NumericDim(2))),
+				new TensorType(FLOAT32, List.of(new NumericDim(2)))), dlSig.get().parameterTypes());
+
+		Function addFunc = nameToFunctions.get("add").iterator().next();
```
[the remark this answers is not in the document]
Comment (khatchad) [reply edu.cuny.hunter.hybridize.tests/test cases/edu/cuny/hunter/hybridize/tests/HybridizeFunctionRefactoringTest.java:2257]: Same as the dense_layer thread above: `nameToFunctions.get("add")` returns a `List`, not a `Set`. Iteration is deterministic, and the outer FUT loop already pins `size() == 1` for every name.

File: edu.cuny.hunter.hybridize.tests/test cases/edu/cuny/hunter/hybridize/tests/HybridizeFunctionRefactoringTest.java [hunk 3]
```diff
@@ -2323,6 +2359,15 @@ public void testHasLikelyTensorParameter15() throws Exception {
 
 			assertTrue("Expecting " + function + " to likely have a tensor-like parameter.", function.getHasTensorParameter());
 		}
+
+		// Precision audit. `double` is called with `tf.constant(1)` (INT32), `tf.constant(1.1)` (FLOAT32), and `tf.constant("a")` /
+		// `tf.constant("b")` (STRING). Multi-context with three distinct dtypes — algorithm drops signature on dtype disagreement
+		// (|D| != 1).
+		Function dbl = nameToFunctions.get("double").iterator().next();
+		Parameter a = dbl.getParameters().get(0);
+		assertTrue("Expected multi-context tensor types from diverging dtypes.", a.getTensorTypes().size() > 1);
```
[the remark this answers is not in the document]
Comment (khatchad) [reply edu.cuny.hunter.hybridize.tests/test cases/edu/cuny/hunter/hybridize/tests/HybridizeFunctionRefactoringTest.java]: Fair point. Strengthened to exact set membership in d5712634: `Set.of(new TensorType(INT32, List.of()), new TensorType(FLOAT32, List.of()), new TensorType(DType.STRING, List.of()))`. A collapse to a single top-type now fails the test.

File: edu.cuny.hunter.hybridize.tests/test cases/edu/cuny/hunter/hybridize/tests/HybridizeFunctionRefactoringTest.java [hunk 4]
```diff
@@ -3740,7 +3777,35 @@ private void testHasLikelyTensorParameterHelper(boolean expectingHybridFunction)
...
+		assertEquals("a", a.getName());
+		Parameter b = params.get(1);
+		assertEquals("b", b.getName());
+
+		assertTrue(function.getHasTensorParameter());
+
+		// Precision audit. `tf.RaggedTensor.from_nested_row_splits(values, splits)` — runtime dtype=int32, shape=(3, None, None).
+		// Ariadne emits an INT32 TensorType with three dims: a known constant 3 followed by two unknown (null) dims, matching the
+		// runtime shape exactly.
+		TensorType ariadne = new TensorType(INT32, java.util.Arrays.asList(new NumericDim(3), null, null));
```
[the remark this answers is not in the document]
Comment (khatchad) [reply edu.cuny.hunter.hybridize.tests/test cases/edu/cuny/hunter/hybridize/tests/HybridizeFunctionRefactoringTest.java:3800]: The `null` entries are what Ariadne actually emits in the shape list for unknown dims today (verified for `tf.RaggedTensor.from_nested_row_splits` — see lines 3796-3801 of the comment block). The test pins current upstream behavior; a helper would reduce to `() -> null` without adding type safety. If Ariadne migrates to a dedicated unknown-dim sentinel, this test fails loudly and we update both sides together. Filing an upstream representation issue is in scope for wala/ML, not this PR.

[the remark this answers is not in the document]
Comment (khatchad) [reply edu.cuny.hunter.hybridize.tests/test cases/edu/cuny/hunter/hybridize/tests/HybridizeFunctionRefactoringTest.java:3800]: Correcting my earlier reply on this thread—the framing was wrong. Per-element `null` is **not** a leak nor an Ariadne-side consistency gap; it is a deliberate (but undocumented and untyped) raggedness marker emitted by Ariadne's ragged generators (`RaggedConstant.java:431`, `RaggedFromNestedRowLengths.java:180`). What was actually brittle—and is the correct audit finding—is the *representation*: ragged is encoded as raw `null` instead of as a typed `RaggedDim` subclass alongside the existing `NumericDim`/`SymbolicDim`/`CompoundDim` hierarchy.

Filed as wala/ML#544 (typed `RaggedDim` sentinel) and #524 (consumer-side `RaggedTensorSpec` emission). Three pinning TODO anchors planted in 9106fd57 so each flip target is visible at the relevant assertion and at the translation site in `Function#inferInputSignature`.
